### DISCUSSION TITLE
Efficient way to build a large community model

### DISCUSSION THREAD

**Discussion by stelmo** [micom-dev/micom#53] [Q&A] [answered]
Hi, I am trying to build a community model and I was wondering what the most efficient way to do this is. Basically, I have a number of metabolic models and I want to join them all together so that they could potentially exchange extracellular metabolites (I want to constrain the extracellular exchange reactions individually in a later analysis). 

I currently have something similar to this implemented, but it is quite slow:
```python
from micom import Community
import os
import pandas as pd

fileloc = os.path.join("data", "models", "ecoli", "iML1515.xml")
ids = ["id"+str(i) for i in range(1, 10, 1)]
df = pd.DataFrame({"id":ids, "file":fileloc, "sample_id":1, "abundance":1}) # assuming equal abundance and only 1 sample with all the microbes

com = Community(df)
``` 
Is there a better/faster way to do this?

Additionally, what kind of information does `com` retain? Does it keep all the annotation information (e.g. gene reaction rules, metabolite/gene/reaction annotations, etc.) for each individual species or does it leave some information out to save on speed/disk space etc.? If this kind of information is kept, is there a way to access the information from `com`?

**Comment by cdiener** [comment, marked as answer]
Not much, since the bottleneck here is the parsing of the model itself from SBML. If you want to build the model repeatedly you could pickle the models beforehand and read those. If the filename ends in `.pickle` this can be used the same way as the SBML model above. The conversion from SBML to the serialized pickle cache can happen in parallel so you could use [multiprocessing.Pool](https://docs.python.org/3/library/multiprocessing.html#using-a-pool-of-workers). Alternatively, you can also build a [MICOM model DB](https://micom-dev.github.io/micom/autoapi/micom/workflows/build/index.html#micom.workflows.build.build_database). This won't be as fast as pickling but will be a bit more standard compliant and connects better to the workflows. Finally, after you built the community model you can save and read it much faster by using 

```Python
com.to_pickle(...)
...
from micom import load_pickle

com = load_pickle(...)
```

If you want to build a lot of community models, and not just the one, [MICOM workflows](https://micom-dev.github.io/micom/high_level.html) can parallelize that for you.

And yes, `com` keeps everything and is just fancy COBRAPY model. So you can still access individual parts like

```
In [1]: import micom as mm

In [2]: com = mm.Community(mm.data.test_taxonomy())
Building ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00

In [3]: com.reactions[0].gene_reaction_rule
Out[3]: 'b0351 or b1241'

In [5]: com.reactions[0].annotation
Out[5]: {'bigg.reaction': 'ACALD'}

In [8]: com.metabolites[0].global_id
Out[8]: 'acald_c'

In [10]: com.metabolites[0].community_id
Out[10]: 'Escherichia_coli_1'

In [11]: com.metabolites[0].annotation
Out[11]: 
{'sbo': 'SBO:0000247',
 'bigg.metabolite': 'acald',
 'ncbigi': '312944817',
 'seed.compound': 'cpd00071',
 'metanetx.chemical': 'MNXM75',
 'kegg.compound': 'C00084',
 'hmdb': 'HMDB00990',
 'chebi': ['CHEBI:13703',
  'CHEBI:15343',
  'CHEBI:22158',
  'CHEBI:2383',
  'CHEBI:40533'],
 'unipathway.compound': 'UPC00084',
 'biocyc': 'META:ACETALD',
 'umbbd.compound': 'c0160',
 'metanetx.reaction': 'MNXR3241'}
```

If you use the workflows the `GrowthResults` object returned there also contains annotations for the exchanges.
